File: src/main/java/pl/training/patterns/behavioral/memento/accounts/Account.java
package pl.training.patterns.behavioral.memento.accounts;

import java.math.BigDecimal;
import java.util.Objects;
import java.util.UUID;

public class Account {

    private final UUID number;
    private BigDecimal balance = BigDecimal.ZERO;

    public Account(final UUID number) {
        this.number = Objects.requireNonNull(number);
    }

    public void deposit(BigDecimal amount) {
        balance = balance.add(amount);
    }

    public Memento createMemento() {
        return new Memento(balance);
    }

    public void restoreMemento(Memento memento) {
        balance = memento.balance();
    }

    public BigDecimal getBalance() {
        return balance;
    }

    @Override
    public String toString() {
        return "Account(number=" + number + ", balance=" + balance + ")";
    }

}

File: csharp/src/Training.Patterns/Behavioral/Memento/Accounts/Account.cs
namespace Training.Patterns.Behavioral.Memento.Accounts;

public class Account
{
    private readonly Guid number;

    public decimal Balance { get; private set; } = decimal.Zero;

    public Account(Guid number)
    {
        this.number = number; // Java: Objects.requireNonNull - Guid is a value type, never null
    }

    public void Deposit(decimal amount)
    {
        Balance += amount;
    }

    public Memento CreateMemento() => new(Balance);

    public void RestoreMemento(Memento memento)
    {
        ArgumentNullException.ThrowIfNull(memento);
        Balance = memento.Balance();
    }

    public override string ToString() => "Account(number=" + number + ", balance=" + Balance + ")";
}
